src/main/scala/se/sics/concepts/core/CGraph.scala

```diff
@@ -0,0 +1,409 @@
+      .map { case (i, (((j, sij), ri), di: Option[Double])) => (j, (i, sij, ri, di.getOrElse(0.0)))}
+      .join(remainingEdgeWeightSums).leftOuterJoin(discardedEdgeWeightSums)
+      // Reformat and fetch optional sums in dj
+      .map { case (j, (((i, sij, ri, di), rj), dj: Option[Double])) => (i, j, sij, ri, rj, di, dj.getOrElse(0.0))}
+  }
+
+
+  /**
+   * Calculates the max relative L1 norm and the max error for each pair.
+   */
+  lazy val relativeL1NormWithErrorBound: RDD[((Long, Long), Double, Double)] = {
+    // Sum terms and normalize with respect to maximum possible L1 norm
+    // The maximum possible error is given by the sum of discarded edge weighs of i and j
+    sharedTermsWithSums
+      .map { case (i, j, sij, ri, rj, di, dj) => {
+        val tot = ri + rj + di + dj
+        ((i, j), (sij + tot) / tot, (di + dj) / tot)
+        }
+      }
+  }
+
+  /** Similarities, where S(i, j) is one subtracted by the relative L1 norm */
+  // TODO: incorporate error bound here as well
+  lazy val simEdges: RDD[((Long, Long), Double)] = {
+    relativeL1NormWithErrorBound.map { case ((i, j), relL1, maxError) => ((i, j), 1.0 - relL1)}
+  }
+
+  def directedSimilarityEdges = simEdges.flatMap{case ((i, j), w) =>  Iterable(((i, j), w), ((j, i), w))}
+
+  // TODO: Duplicates some code here - implement separate index to label function instead
+  lazy val relativeL1NormWithLabels: RDD[((String, String), Double, Double)] = {
+    // Key-value pairs from indices to strings
+    val vertexIndexStrings = prunedVertices.map{case (i, si, wi) => (i, si)}
+    // Map indices to labels
+    relativeL1NormWithErrorBound
+      .map{case ((i, j), wij, d) => (i, (j, wij, d))}
+      .join(vertexIndexStrings)
+      .map{case (i, ((j, wij, d), si)) => (j, (si, wij, d))}
+      .join(vertexIndexStrings)
+      .map{case (j, ((si, wij, d), sj)) => ((si, sj), wij, d)}
+  }
+
+  // Key-value pairs from indices to strings
+  lazy val vertexIndexStrings = prunedVertices.map { case (i, si, wi) => (i, si) }
+
+  // Similarities with labels
+  lazy val similarityEdgesWithLabels: RDD[((String, String), Double)] = {
+    // Map indices to labels
+    simEdges
+      .map { case ((i, j), wij) => (i, (j, wij)) }
+      .join(vertexIndexStrings)
+      .map { case (i, ((j, wij), si)) => (j, (si, wij)) }
+      .join(vertexIndexStrings)
+      .map { case (j, ((si, wij), sj)) => ((si, sj), wij) }
+  }
+
+  // Test of community detection using GraphX' label propagation algorithm
+  // Note: Current GraphX implementation does not take edge weights into consideration
+  def clustersBySimilarityGraphX(numberOfIterations: Int): RDD[(Long, Long)] = {
+    val edges: RDD[Edge[Double]] = directedSimilarityEdges.map{case ((i, j), w) => Edge(i, j, w)}
+    val similarityGraph: Graph[String, Double] = Graph(vertexIndexStrings, edges)
+    LabelPropagation.run(similarityGraph, numberOfIterations).vertices
+  }
+
+  def clustersByConnectivityGraphX(): RDD[(Long, Long)] = {
```

Comment:
Removing this would mean we do no clustering for FigureOne example. Is this what we want?